### ISSUE TITLE
Value tracking tool window needs more/better information

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#59621]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Value-tracking-tool-window-needs-morebe/1655832)._

---
The current layout and presentation of the "Value Tracking" tool is confusing and insufficient.
After selecting a C# variable in the text editor and then "Track Value Source" a window pops up showing seemingly unrelevant information somehow relating to the selected variable. 
The window is split in two parts, the left one showing occurances (or something like that) and the right one simply a file name and line number of the selected entry in the left part.
I would expect at least a code preview on the right hand part.
As for the left hand part, the color scheme makes no sense, there are variables highlighted without any relation, lines appear multiples times with only varying icons and no further explanation, no tooltips of any kinds are given.
As its called "Value Tracking" one expects to see a "timeline"- or "callstack"-like display of the value and its history through the code. This is definitely not what one gets from this tool yet.
There is also seemingly no documentation at all available from the VS doc site.

! [image.png] (https://aka.ms/dc/image?name=B5adde864f91642d5b226ece246e6977a637798318005000405_20220207-115628-image.png&tid=5adde864f91642d5b226ece246e6977a637798318005000405)

---
#### Feedback Bot on 2/7/2022, 06:54 PM: 

Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further. You will hear from us in about a week on our next steps.

**Comment by jcouv** [comment]
FYI @mikadumont

**Comment by ericsampson** [comment]
I am so so excited about data flow analysis for code understanding being this easy, but currently the Ux of the tool is so alien IMO as to be unusable.

Translating from the code to the UI presented in this simple example is super unintuitive, never mind in a real codebase.

![image](https://user-images.githubusercontent.com/1183853/234958225-1f7a6b9f-d8d4-4bd2-9fe0-e671499deaa8.png)

**Comment by CyrusNajmabadi** [comment]
@ericsampson PRs welcome :)

**Comment by ryzngard** [comment]
@ericsampson agree this has some pain points that could be improved. In the immediate above, I see a few things: 

1. Adding a preview on the split window sounds neat. I don't think that would be terribly hard (but I've never worked with the preview window)
2. Fixing ordering: a sort of easy fix would be to try and order by file and line for each level in the tree. That would put `startingString += i` above the `someHighLevelMethod` call.
3. Reducing noise: the `return someLowLevelMethod([|theString]])` isn't needed here. It's included because it's a symbolic reference, but I think we can be smarter about that. 

As @CyrusNajmabadi said PRs are welcome. I am happy to help point you in the right direction.

**Comment by ericsampson** [comment]
@CyrusNajmabadi I definitely didn't intend to come across combative, if so I apologize it's been an extremely stressful week at work.

I'll do a little thinking to see what end result visualization might bring clarity, maybe post a mockup screenshot.

I'd definitely be interested in seeing where the code would be to tweak @ryzngard if it's not a monumental Roslyn newbie task :)

Cheers!

**Comment by ericsampson** [comment]
And again, I am incredibly excited by the addition of this feature!!!!
also the Call Hierarchy feature. They're both amazing

**Comment by ryzngard** [comment]
@ericsampson the endpoint that produces the tree is https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/ValueTracking/ValueTracker.cs#L26

The UI portion for VS will be in https://github.com/dotnet/roslyn/tree/main/src/VisualStudio/Core/Def/ValueTracking 

I'd suggest picking one of the things above to be pointed about your exploration or anything tagged with the label I added here. We also have a Roslyn discord channel http://aka.ms/discord-csharp-roslyn . We have gitter but admittedly I don't use that :/

**Comment by ryzngard** [comment]
> I definitely didn't intend to come across combative, if so I apologize it's been an extremely stressful week at work.

Also, on this, I think you came across fine and just wanting to see improvement. I wrote the feature, and it's good to see people wanting to use it and improve it. Cyrus does a great job of reminding people we're open source to empower them to make the changes they want  :)

**Comment by CyrusNajmabadi** [comment]
I didn't think you were combative.  I just like fishing to see if passionate people are interested in working on pet-peeve areas :)

**Comment by ericsampson** [comment]
Thank you both, I'm excited to see if I can get some time to experiment!

**Comment by ericsampson** [comment]
@ryzngard is the "trying your changes in VS" part of this doc the dev workflow is do for making changes to the view?

https://github.com/dotnet/roslyn/blob/main/docs/contributing/Building,%20Debugging,%20and%20Testing%20on%20Windows.md

**Comment by ryzngard** [comment]
> @ryzngard is the "trying your changes in VS" part of this doc the dev workflow is do for making changes to the view?
> 
> https://github.com/dotnet/roslyn/blob/main/docs/contributing/Building,%20Debugging,%20and%20Testing%20on%20Windows.md

Correct. In general if you have the requirements met it should be as easy as f5 deploy. I heavily recommend "just my code" for debug settings
